src/H5private.h

```diff
@@ -1562,6 +1562,25 @@ H5_DLL int HDvasprintf(char **bufp, const char *fmt, va_list _ap);
 #define H5_CHECK_OVERFLOW(var, vartype, casttype)
 #endif /* NDEBUG */
 
+/*
+ * A macro for checking whether a multiplication has overflown
+ * r is assumed to be the result of a prior multiplication of a and b
+ */
+#define H5_CHECK_MUL_OVERFLOW(r, a, b, err)                                                                  \
```

Comment:
Shouldn't this macro be surrounded by "#ifndef NDEBUG" like the rest of the macros?
If not then maybe it needs to be a different kind of macro - do we have a section for runtime NDEBUG type macros?

Reply:
As for the macro location, if there is a better place to put it, let me know, I'd be happy to move it there. I've created a macro instead of coding it directly as I assumed it may be useful in other places as well.

Reply:
I don't know the answer, I just noticed something not like the others. We will need a library dev to answer.

Reply:
I suggest a fix in place in H5O__attr_decode() of H5Oattr.c as follows:

+    if(ds_size != 0 && ((attr->shared->data_size/ds_size) != dt_size))
+        HGOTO_ERROR(H5E_RESOURCE, H5E_OVERFLOW, NULL, "data size exceeds addressable range")
+
    
Note that dt_size cannot be zero, if so, it will flag as error few lines back:
  
    if (0 == (dt_size = H5T_get_size(attr->shared->dt)))
        HGOTO_ERROR(H5E_ATTR, H5E_CANTGET, NULL, "unable to get datatype size")

With the above fix, h5dump output on the corrupted file will be something like this:

 GROUP "input_1" {h5dump error: error getting attribute information

   }

I also suggest to add a test to verify the fix with the corrupted file.

Reply:
I had created a more generic macro in the hope it may be useful in other cases as well. But it can certainly be done explicitly. If we do so and since we already know that `dt_sizew` != 0, we could take it as the divisor and simply do:
``` 
if ((attr->shared->data_size/dt_size) != ds_size)
       HGOTO_ERROR(...)
```